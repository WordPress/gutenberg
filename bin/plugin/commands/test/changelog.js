/**
 * Internal dependencies
 */
import {
	getNormalizedTitle,
	reword,
	addTrailingPeriod,
	createOmitByTitlePrefix,
	createOmitByLabel,
	capitalizeAfterColonSeparatedPrefix,
	getIssueType,
	sortGroup,
	skipCreatedByBots,
	getTypesByLabels,
	getTypesByTitle,
	getIssueFeature,
	getFormattedItemDescription,
	getUniqueByUsername,
	getChangelog,
	getContributorProps,
	getContributorsList,
	mapLabelsToFeatures,
} from '../changelog';
import _pullRequests from './fixtures/pull-requests.json';
import botPullRequestFixture from './fixtures/bot-pull-requests.json';

/**
 * pull-requests.json is a static snapshot of real data from the GitHub API.
 * We merge this with dummy fixture data for a "bot" pull request so as to
 * ensure future updates to the pull-requests.json doesn't reduce test coverage
 * of filtering out of bot PRs.
 * See: https://github.com/WordPress/gutenberg/pull/38777#discussion_r808992346.
 */
const pullRequests = _pullRequests.concat( botPullRequestFixture );

describe( 'getNormalizedTitle', () => {
	const DEFAULT_ISSUE = {
		labels: [],
	};

	it.each( [
		[ 'adds period', 'Fixes a bug', 'Fixes a bug.' ],
		[ 'keeps period', 'Fixes a bug.', 'Fixes a bug.' ],
		[
			'omits mobile by title',
			'[RNMoBILe] Address mobile concern',
			undefined,
		],
		[
			'omits mobile by issue',
			'Address mobile concern',
			undefined,
			{
				...DEFAULT_ISSUE,
				labels: [ { name: 'Mobile App - i.e. Android or iOS' } ],
			},
		],
		[
			'avoids reword of joined terms',
			'e2e-tests: Improve test stability',
			'e2e-tests: Improve test stability.',
		],
		[
			'rewords',
			'Improve e2e url stability',
			'Improve end-to-end URL stability.',
		],
		[ 'capitalizes', 'fix bug', 'Fix bug.' ],
		[
			'removes redundant prefix',
			'Code quality: Enable import/no-unresolved ESLint rule for Gutenberg',
			'Enable import/no-unresolved ESLint rule for Gutenberg.',
			{
				...DEFAULT_ISSUE,
				labels: [ { name: '[Type] Code Quality' } ],
			},
		],
		[
			'removes redundant prefix with normalized type',
			'[Enhancement] Add ability to transform audio shortcodes to audio blocks',
			'Add ability to transform audio shortcodes to audio blocks.',
			{
				...DEFAULT_ISSUE,
				labels: [ { name: '[Type] Enhancement' } ],
			},
		],
	] )( '%s', ( _label, original, expected, issue = DEFAULT_ISSUE ) => {
		expect( getNormalizedTitle( original, issue ) ).toBe( expected );
	} );
} );

describe( 'addTrailingPeriod', () => {
	it( 'adds a period if missing', () => {
		const result = addTrailingPeriod( 'Fixes a bug' );

		expect( result ).toBe( 'Fixes a bug.' );
	} );

	it( 'does not add a period if already present', () => {
		const result = addTrailingPeriod( 'Fixes a bug.' );

		expect( result ).toBe( 'Fixes a bug.' );
	} );

	it( 'trims trailing whitespace before appending period', () => {
		const result = addTrailingPeriod( 'Fixes a bug ' );

		expect( result ).toBe( 'Fixes a bug.' );
	} );
} );

describe( 'createOmitByTitlePrefix', () => {
	it( 'returns identity if not containing matching prefix', () => {
		const result = createOmitByTitlePrefix( [ '[omIT]' ] )( 'Fix' );

		expect( result ).toBe( 'Fix' );
	} );

	it( 'returns undefined if given prefix', () => {
		const result = createOmitByTitlePrefix( [ '[omIT]' ] )( '[omit] Fix' );

		expect( result ).toBe( undefined );
	} );
} );

describe( 'createOmitByLabel', () => {
	it( 'returns identity if label is not assigned to issue', () => {
		const result = createOmitByLabel( [ 'Omit' ] )( 'Fix', { labels: [] } );

		expect( result ).toBe( 'Fix' );
	} );

	it( 'returns undefined if given prefix', () => {
		const result = createOmitByLabel( [ 'Omit' ] )( 'Fix', {
			labels: [ { name: 'Omit' } ],
		} );

		expect( result ).toBe( undefined );
	} );
} );

describe( 'reword', () => {
	it( 'avoids reword of joined terms', () => {
		const result = reword( 'e2e-tests: Improve test stability' );

		expect( result ).toBe( 'e2e-tests: Improve test stability' );
	} );

	it( 'rewords terms', () => {
		const result = reword( 'Improve e2e url stability' );

		expect( result ).toBe( 'Improve end-to-end URL stability' );
	} );
} );

describe( 'capitalizeAfterColonSeparatedPrefix', () => {
	it( 'capitalizes the last segment after a colon', () => {
		const result = capitalizeAfterColonSeparatedPrefix( 'blocks: fix bug' );

		expect( result ).toBe( 'blocks: Fix bug' );
	} );
} );

describe( 'getIssueType', () => {
	it( 'returns various if unable to find appropriate type by label', () => {
		const result = getIssueType( { labels: [] } );

		expect( result ).toBe( 'Various' );
	} );

	it( 'returns type by label', () => {
		const result = getIssueType( {
			labels: [ { name: '[Type] Code Quality' } ],
		} );

		expect( result ).toBe( 'Code Quality' );
	} );

	it( 'returns remapped type by label', () => {
		const result = getIssueType( { labels: [ { name: '[Type] Bug' } ] } );

		expect( result ).toBe( 'Bug Fixes' );
	} );

	it( 'prioritizes by group order', () => {
		const result = getIssueType( {
			labels: [ { name: '[Type] Task' }, { name: '[Type] Enhancement' } ],
		} );

		expect( result ).toBe( 'Enhancements' );
	} );

	it( 'prioritizes meta categories', () => {
		const result = getIssueType( {
			labels: [
				{ name: '[Type] Bug' },
				{ name: '[Type] Build Tooling' },
			],
		} );

		expect( result ).toBe( 'Tools' );
	} );
} );

describe( 'getIssueFeature', () => {
	it( 'returns "Unknown" as feature if there are no labels', () => {
		const result = getIssueFeature( { labels: [] } );

		expect( result ).toBe( 'Uncategorized' );
	} );

	it( 'falls by to "Unknown" as the feature if unable to classify by other means', () => {
		const result = getIssueFeature( {
			labels: [
				{
					name: 'Some Label',
				},
				{
					name: '[Package] Example Package', // 1. has explicit mapping.
				},
				{
					name: '[Package] Another One',
				},
			],
		} );

		expect( result ).toEqual( 'Uncategorized' );
	} );

	it( 'gives precedence to manual feature mapping', () => {
		const result = getIssueFeature( {
			labels: [
				{
					name: '[Block] Some Block', // 3. Block-specific label.
				},
				{
					name: '[Package] Edit Widgets', // 1. has explicit mapping.
				},
				{
					name: '[Feature] Some Feature', // 2. Feature label.
				},
				{
					name: '[Package] Another One',
				},
			],
		} );

		const mappingForPackageEditWidgets = 'Widgets Editor';

		expect( result ).toEqual( mappingForPackageEditWidgets );
	} );

	it( 'gives secondary priority to feature labels when manually mapped label is not present', () => {
		const result = getIssueFeature( {
			labels: [
				{
					name: '[Block] Some Block', // Block specific label.
				},
				{
					name: '[Package] This package',
				},
				{
					name: '[Feature] Cool Feature', // Should have priority despite prescence of block specific label.
				},
				{
					name: '[Package] Another One',
				},
			],
		} );

		expect( result ).toEqual( 'Cool Feature' );
	} );

	it( 'gives tertiary priority to "Block Library" as feature for all PRs that have a block specific label (and where manually mapped or feature label not present)', () => {
		const result = getIssueFeature( {
			labels: [
				{
					name: '[Block] Some Block',
				},
				{
					name: '[Package] This package',
				},
				{
					name: '[Package] Another One',
				},
			],
		} );

		expect( result ).toEqual( 'Block Library' );
	} );
} );

describe( 'sortGroup', () => {
	it( 'returns groups in order', () => {
		const result = [
			'Code Quality',
			'Bug Fixes',
			'Various',
			'New APIs',
			'Enhancements',
			'Performance',
		].sort( sortGroup );

		expect( result ).toEqual( [
			'Enhancements',
			'New APIs',
			'Bug Fixes',
			'Performance',
			'Code Quality',
			'Various',
		] );
	} );
} );

describe( 'getTypesByLabels', () => {
	it( 'returns all normalized type candidates by type prefix. it is case insensitive', () => {
		const result = getTypesByLabels( [
			'[Type] Regression',
			'[Type] Bug',
			'[Package] Blocks',
			'[Type] performance',
		] );

		expect( result ).toEqual( [ 'Bug Fixes', 'Performance' ] );
	} );
} );

describe( 'mapLabelsToFeatures', () => {
	it( 'returns all normalized feature candidates by feature prefix. it is case insensitive', () => {
		const result = mapLabelsToFeatures( [
			'[Package] Commands',
			'[Package] Block Library',
			'[Feature] Link Editing',
			'[Feature] block Multi Selection',
		] );

		expect( result ).toEqual( [
			'Commands',
			'Block Library',
			'Block Editor',
		] );
	} );
} );

describe( 'getTypesByTitle', () => {
	it.each( [
		[ 'Fix Typography panel rendering from style hooks' ],
		[ 'fix: unset max-width for reusable blocks' ],
		[
			'Bug fix "Cannot read property \'end\' of undefined" on babel-plugin-makepot #21466',
		],
		[ 'Editor: Fix "Attempt Recovery" error boundary handler' ],
		[ 'Fix/Remove edit gallery from media library modal' ],
		[ 'Fixes a broken dev doc example for plugin Sidebars' ],
	] )( 'returns bug type by title (%s)', ( title ) => {
		const result = getTypesByTitle( title );

		expect( result ).toEqual( [ 'Bug Fixes' ] );
	} );
} );

describe( 'getUniqueByUsername', () => {
	it( 'removes duplicate entries by username', () => {
		const entries = [
			{
				user: {
					login: '@user1',
				},
			},
			{
				user: {
					login: '@user1',
				},
			},
			{
				user: {
					login: '@user2',
				},
			},
			{
				user: {
					login: '@user3',
				},
			},
			{
				user: {
					login: '@user4',
				},
			},
		];

		const expected = [
			{
				user: {
					login: '@user1',
				},
			},
			{
				user: {
					login: '@user2',
				},
			},
			{
				user: {
					login: '@user3',
				},
			},
			{
				user: {
					login: '@user4',
				},
			},
		];
		expect( getUniqueByUsername( entries ) ).toEqual( expected );
	} );
} );

describe( 'skipCreatedByBots', () => {
	it( 'removes entries created by bots', () => {
		const entries = [
			{
				user: {
					login: '@user1',
					type: 'User',
				},
			},
			{
				user: {
					login: '@dependabot[bot]',
					type: 'Bot',
				},
			},
			{
				user: {
					login: '@user2',
					type: 'User',
				},
			},
			{
				user: {
					login: '@someotherrandombotusername',
					type: 'Bot',
				},
			},
			{
				user: {
					login: '@user3',
					type: 'User',
				},
			},
		];

		const expected = [
			{
				user: {
					login: '@user1',
					type: 'User',
				},
			},
			{
				user: {
					login: '@user2',
					type: 'User',
				},
			},
			{
				user: {
					login: '@user3',
					type: 'User',
				},
			},
		];
		expect( skipCreatedByBots( entries ) ).toEqual( expected );
	} );
} );

describe( 'getFormattedItemDescription', () => {
	it( 'creates a markdown formatted description', () => {
		const expected =
			'This is a test title and should have a link. ([123456](https://github.com/123456))';
		expect(
			getFormattedItemDescription(
				'This is a test title and should have a link.',
				123456,
				'https://github.com/123456'
			)
		).toEqual( expected );
	} );
} );

describe( 'getChangelog', () => {
	test( 'verify that the changelog is properly formatted', () => {
		// The fixture with the list of pull requests was generated by running the following command:
		// npm run other:changelog -- --milestone="Gutenberg 16.8"
		// The response from the `fetchAllPullRequests` call in the `getChangelog` method was stored in the JSON file.
		expect( getChangelog( pullRequests ) ).toMatchSnapshot();
	} );
} );

describe( 'getContributorProps', () => {
	test( 'verify that the contributors props are properly formatted', () => {
		// The fixture with the list of pull requests was generated by running the following command:
		// npm run other:changelog -- --milestone="Gutenberg 11.3"
		expect( getContributorProps( pullRequests ) ).toMatchSnapshot();
	} );
	test( 'do not include first time contributors section if there are not any', () => {
		expect(
			getContributorProps( pullRequests.slice( 0, 4 ) )
		).toMatchInlineSnapshot( `""` );
	} );
} );

describe( 'getContributorList', () => {
	test( 'verify that the contributors list is properly formatted', () => {
		// The fixture with the list of pull requests was generated by running the following command:
		// npm run other:changelog -- --milestone="Gutenberg 11.3"
		expect( getContributorsList( pullRequests ) ).toMatchSnapshot();
	} );
} );
