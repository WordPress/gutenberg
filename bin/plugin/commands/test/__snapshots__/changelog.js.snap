// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getChangelog verify that the changelog is properly formatted 1`] = `
"## Changelog

### Features

#### Block Editor
- Adds 'nofollow' setting to Button block. ([54110](https://github.com/WordPress/gutenberg/pull/54110))

#### Site Editor
- Add 'Show template' toggle when editing pages. ([52674](https://github.com/WordPress/gutenberg/pull/52674))


### Enhancements

#### Components
- Adding label/description to \`BlockEditor/DuotoneControl\`. ([54473](https://github.com/WordPress/gutenberg/pull/54473))
- Deprecating \`isPressed\` in \`Button\` component. ([54740](https://github.com/WordPress/gutenberg/pull/54740))
- InputControl-based components: Add opt-in prop for next 40px default size. ([53819](https://github.com/WordPress/gutenberg/pull/53819))
- Modal: Add \`contentWidth\` prop to support a selection of preset modal sizes. ([54471](https://github.com/WordPress/gutenberg/pull/54471))
- Remove unused components from \`ui/\`. ([54573](https://github.com/WordPress/gutenberg/pull/54573))
- Update compact search control metrics. ([54663](https://github.com/WordPress/gutenberg/pull/54663))
- Wrapped \`TextareaControl\` in a \`forwardRef\` call. ([54975](https://github.com/WordPress/gutenberg/pull/54975))

#### Block Library
- Add a brief description to the Footnotes block. ([54613](https://github.com/WordPress/gutenberg/pull/54613))
- Footnotes: Use core’s meta revisioning if available. ([52988](https://github.com/WordPress/gutenberg/pull/52988))
- Login/out: Add spacing support. ([45147](https://github.com/WordPress/gutenberg/pull/45147))
- Query view.js: Code quality. ([54982](https://github.com/WordPress/gutenberg/pull/54982))
- Set custom color when applying initial background image. ([54054](https://github.com/WordPress/gutenberg/pull/54054))
- Use \`wp_get_inline_script_tag()\` in \`build_dropdown_script_block_core_categories()\`. ([54637](https://github.com/WordPress/gutenberg/pull/54637))

#### Block Editor
- Default suggested links to pages. ([54622](https://github.com/WordPress/gutenberg/pull/54622))
- Remove base URL from link control search results. ([54553](https://github.com/WordPress/gutenberg/pull/54553))
- Simplify \`BlockHTMLConvertButton\`. ([54972](https://github.com/WordPress/gutenberg/pull/54972))
- Update strings in blocks 'RenameModal' component. ([54887](https://github.com/WordPress/gutenberg/pull/54887))

#### Post Editor
- Edit Post: Use hooks instead of HoCs in 'PostStatus' components. ([54951](https://github.com/WordPress/gutenberg/pull/54951))
- Editor: Use hooks instead of HoCs in 'PostSticky' components. ([54949](https://github.com/WordPress/gutenberg/pull/54949))
- Editor: Use hooks instead of HoCs in 'PostSwitchToDraftButton'. ([54695](https://github.com/WordPress/gutenberg/pull/54695))
- Show confirmation dialog when moving a post to the trash. ([50219](https://github.com/WordPress/gutenberg/pull/50219))

#### Patterns
- Remove category description in inserter panel. ([54894](https://github.com/WordPress/gutenberg/pull/54894))

#### Typography
- Font Library: Refactor endpoint permissions. ([54829](https://github.com/WordPress/gutenberg/pull/54829))

#### Site Editor
- Add template replace flow to template inspector. ([54609](https://github.com/WordPress/gutenberg/pull/54609))


### Bug Fixes

- Fix the ShortcutProvider usage. ([54851](https://github.com/WordPress/gutenberg/pull/54851))
- Fix warning when a template calls a template area twice. ([54861](https://github.com/WordPress/gutenberg/pull/54861))
- Revert "Fix warning when a template calls a template area twice". ([54926](https://github.com/WordPress/gutenberg/pull/54926))

#### Block Library
- All Nav block items to break long titles. ([54866](https://github.com/WordPress/gutenberg/pull/54866))
- Fallback to Twitter provider when embedding X URLs. ([54876](https://github.com/WordPress/gutenberg/pull/54876))
- Fix Deleted Navigation Menu warning string. ([55033](https://github.com/WordPress/gutenberg/pull/55033))
- Fix Search Block not updating in Nav block. ([54823](https://github.com/WordPress/gutenberg/pull/54823))
- Fix left and right aligmnent in children of Post Template. ([54997](https://github.com/WordPress/gutenberg/pull/54997))
- Fix output of Navigation block classnames in the editor. ([54992](https://github.com/WordPress/gutenberg/pull/54992))
- Fix overwriting of published post meta when previewing footnote changes. ([54339](https://github.com/WordPress/gutenberg/pull/54339))
- Image: Ensure Expand on Click toggle is shown if block-level lightbox setting exists. ([54878](https://github.com/WordPress/gutenberg/pull/54878))
- Image: Fix layout shift when lightbox is opened and closed. ([53026](https://github.com/WordPress/gutenberg/pull/53026))
- Media & Text: Fix React warning. ([55038](https://github.com/WordPress/gutenberg/pull/55038))
- Search block: Allow space for input field only when form expanded. ([54846](https://github.com/WordPress/gutenberg/pull/54846))
- Search block: Update alignment and icon button width. ([54773](https://github.com/WordPress/gutenberg/pull/54773))
- Update pattern import menu item. ([54782](https://github.com/WordPress/gutenberg/pull/54782))

#### Site Editor
- Avoid same key warnings in template parts area listings. ([54863](https://github.com/WordPress/gutenberg/pull/54863))
- Avoid stale navigation block values when parsing entity record. ([54996](https://github.com/WordPress/gutenberg/pull/54996))
- Don't display the navigation section in template parts details when a menu is missing. ([54993](https://github.com/WordPress/gutenberg/pull/54993))
- Fix ToolSelector popover variant. ([54840](https://github.com/WordPress/gutenberg/pull/54840))
- Reset 'Show template' toggle when leaving edit mode. ([54679](https://github.com/WordPress/gutenberg/pull/54679))
- [Site Editor]: Update copy of using  the default template in a page. ([54728](https://github.com/WordPress/gutenberg/pull/54728))
- remove \`overflow: Hidden\` from the entity title in the site editor sidebar. ([54769](https://github.com/WordPress/gutenberg/pull/54769))

#### Components
- FormTokenField: Add \`box-sizing\` reset style and reset default padding. ([54734](https://github.com/WordPress/gutenberg/pull/54734))
- Popover: Fix the styles for components that use emotion within popovers. ([54912](https://github.com/WordPress/gutenberg/pull/54912))
- Remove hover style for secondary Button when aria-disabled is set. ([54978](https://github.com/WordPress/gutenberg/pull/54978))
- Reverting addition of \`aria-selected\` style hook in \`Button\`. ([54931](https://github.com/WordPress/gutenberg/pull/54931))
- \`SlotFill\`: Pass \`Component\` instance to unregisterSlot. ([54765](https://github.com/WordPress/gutenberg/pull/54765))

#### Block Editor
- Avoid double-wrapping selectors when transforming the styles. ([54981](https://github.com/WordPress/gutenberg/pull/54981))
- [Inserter]: Fix reset of registered media categories. ([55012](https://github.com/WordPress/gutenberg/pull/55012))

#### Typography
- Font Library: Changed the OTF mime type expected value to be what PHP returns. ([54886](https://github.com/WordPress/gutenberg/pull/54886))
- Font Library: Move font uploads to a new tab. ([54655](https://github.com/WordPress/gutenberg/pull/54655))

#### Global Styles
- Block custom CSS: Fix incorrect CSS when multiple root selectors. ([53602](https://github.com/WordPress/gutenberg/pull/53602))
- Image: Ensure \`false\` values are preserved in memory when defined in \`theme.json\`. ([54639](https://github.com/WordPress/gutenberg/pull/54639))

#### List View
- Fix performance issue when selecting all blocks. ([54900](https://github.com/WordPress/gutenberg/pull/54900))

#### Colors
- Format Library: Try to fix highlight popover jumping. ([54736](https://github.com/WordPress/gutenberg/pull/54736))

#### Interactivity API
- Image: Fix duotone not being applied to lightbox image. ([54670](https://github.com/WordPress/gutenberg/pull/54670))


### Accessibility

#### Components
- Follow ariakit best practices. ([54696](https://github.com/WordPress/gutenberg/pull/54696))
- HTML block: Fix accessibility issues on back-end. ([54408](https://github.com/WordPress/gutenberg/pull/54408))
- Update ariakit to 0.3.3. ([54818](https://github.com/WordPress/gutenberg/pull/54818))
- \`Modal\`: Accessibly hide/show outer modal when nested. ([54743](https://github.com/WordPress/gutenberg/pull/54743))

#### Block Library
- Footnotes: Add aria-label to return links. ([54843](https://github.com/WordPress/gutenberg/pull/54843))
- Table of contents block accessibility improvements. ([54322](https://github.com/WordPress/gutenberg/pull/54322))

#### Patterns
- Use list role instead of listbox in patterns list. ([54884](https://github.com/WordPress/gutenberg/pull/54884))

#### Post Editor
- Editor: Always render the 'Switch to Draft' button to avoid focus loss. ([54722](https://github.com/WordPress/gutenberg/pull/54722))

#### Block Editor
- Block Switcher: Use a different label for multi-selection. ([54692](https://github.com/WordPress/gutenberg/pull/54692))


### Performance

- Tests: Support the Site Editor's legacy spinner. ([54784](https://github.com/WordPress/gutenberg/pull/54784))
- Use instanceOf over property_exists. ([54835](https://github.com/WordPress/gutenberg/pull/54835))

#### Block Editor
- Subscribe only to block editor store in \`useBlockSync\`. ([55041](https://github.com/WordPress/gutenberg/pull/55041))


### Experiments

#### Site Editor
- al]: First version of pages list in site editor. ([54966](https://github.com/WordPress/gutenberg/pull/54966))

#### Block Editor
- Expose \`getDuotoneFilter()\` as private API. ([54905](https://github.com/WordPress/gutenberg/pull/54905))


### Documentation

- Add a documentation page about the block editor settings. ([54870](https://github.com/WordPress/gutenberg/pull/54870))
- Add a page about the format library to the platform documentation site. ([55037](https://github.com/WordPress/gutenberg/pull/55037))
- Docs: Add a callout to the \`wp-now\` documentation to mention it's still experimental. ([55044](https://github.com/WordPress/gutenberg/pull/55044))
- Docs: Remove outdated info. ([54707](https://github.com/WordPress/gutenberg/pull/54707))
- Docs: Remove the Full Site Editing doc. ([54516](https://github.com/WordPress/gutenberg/pull/54516))
- Docs: Rename Block Hooks handbook page to Block Filters. ([54862](https://github.com/WordPress/gutenberg/pull/54862))
- Document the current state of the Real-Time collaboration experiment. ([54932](https://github.com/WordPress/gutenberg/pull/54932))
- Fix a broken MD link in callout. ([54772](https://github.com/WordPress/gutenberg/pull/54772))
- Platform Docs: Add a documentation page explaining how to use the block library. ([54967](https://github.com/WordPress/gutenberg/pull/54967))
- Update the documentation of the block editor and replace @wordpress/element with react. ([54908](https://github.com/WordPress/gutenberg/pull/54908))
- Update versions in WP for 6.4. ([54890](https://github.com/WordPress/gutenberg/pull/54890))


### Code Quality

- Add a unit test for the "ValidBlockLibraryFunctionNameSniff" sniff. ([53928](https://github.com/WordPress/gutenberg/pull/53928))
- Move dependencies to the right place. ([54597](https://github.com/WordPress/gutenberg/pull/54597))
- Move mime-type collection generation to a function that can be tested…. ([54844](https://github.com/WordPress/gutenberg/pull/54844))
- Post Title block should use esc_url(). ([53981](https://github.com/WordPress/gutenberg/pull/53981))
- Rich text: Use getPasteEventData. ([55048](https://github.com/WordPress/gutenberg/pull/55048))
- Writing flow: Absorb clipboard handler. ([55006](https://github.com/WordPress/gutenberg/pull/55006))

#### Block Library
- Footnotes: Avoid regexes in entity provider. ([54505](https://github.com/WordPress/gutenberg/pull/54505))
- Image Block: Fix browser console error when clicking "Expand on Click". ([54938](https://github.com/WordPress/gutenberg/pull/54938))
- Removed unwanted space from the string. ([54654](https://github.com/WordPress/gutenberg/pull/54654))
- Update CODEOWNERS for \`core/image\` block. ([54793](https://github.com/WordPress/gutenberg/pull/54793))

#### Patterns
- Add a new spec for for adding an unsynced pattern. ([54892](https://github.com/WordPress/gutenberg/pull/54892))
- Add end-to-end tests for filtering and searching patterns. ([54906](https://github.com/WordPress/gutenberg/pull/54906))
- Add new end-to-end test for creating a pattern. ([54855](https://github.com/WordPress/gutenberg/pull/54855))
- Include pattern category in main end-to-end critical path test. ([54923](https://github.com/WordPress/gutenberg/pull/54923))

#### Components
- Consolidate utils to remove \`ui/\`. ([54922](https://github.com/WordPress/gutenberg/pull/54922))
- Move \`ContextSystemProvider\` out of \`/ui\`. ([54847](https://github.com/WordPress/gutenberg/pull/54847))
- SlotFill: Migrate to Typescript. ([51350](https://github.com/WordPress/gutenberg/pull/51350))
- Tidying \`CircularOptionPicker.Option\`. ([54903](https://github.com/WordPress/gutenberg/pull/54903))

#### Typography
- Font Library: Syntax refactor repace strpos with str_contains. ([54832](https://github.com/WordPress/gutenberg/pull/54832))
- Font Library: Use snake_case instead of camelCase on fontFamilies endpoint param. ([54977](https://github.com/WordPress/gutenberg/pull/54977))

#### Block Editor
- Rich text: Avoid shortcode logic, adjust paste handler instead. ([55052](https://github.com/WordPress/gutenberg/pull/55052))

#### Plugin
- Remove legacy logic for '__unstableResolvedAssets' setting. ([54812](https://github.com/WordPress/gutenberg/pull/54812))

#### Data Layer
- createResolversCacheMiddleware: Remove dependency on core/data store. ([54733](https://github.com/WordPress/gutenberg/pull/54733))

#### Site Editor
- Use constants rather than hard coded template strings (round 3). ([54705](https://github.com/WordPress/gutenberg/pull/54705))


### Tools

- Label enforcer workflow: Make accessibility a focus instead of a type. ([54941](https://github.com/WordPress/gutenberg/pull/54941))
- Scripts: Update webpack and related dependencies to the latest version. ([54657](https://github.com/WordPress/gutenberg/pull/54657))
- Update changelog automation and test fixtures to match the last a11y label renaming. ([54974](https://github.com/WordPress/gutenberg/pull/54974))

#### Testing
- Don’t use TypeScript files in scripts package. ([54856](https://github.com/WordPress/gutenberg/pull/54856))
- ESLint: Update eslint-plugin-testing-library to v6. ([54910](https://github.com/WordPress/gutenberg/pull/54910))
- Fix end-to-end test: ”WP Editor Meta Boxes > Should save the changes”. ([51884](https://github.com/WordPress/gutenberg/pull/51884))
- Font Library: Avoid deprected error in test. ([54802](https://github.com/WordPress/gutenberg/pull/54802))
- Make \`editor.getBlocks\` to return only testing-related properties. ([54901](https://github.com/WordPress/gutenberg/pull/54901))
- Migrate 'Global styles sidebar' test to Playwright. ([55045](https://github.com/WordPress/gutenberg/pull/55045))
- Migrate 'iframed block editor settings styles' tests to Playwright. ([55014](https://github.com/WordPress/gutenberg/pull/55014))
- Migrate 'iframed inline styles' tests to Playwright. ([55009](https://github.com/WordPress/gutenberg/pull/55009))
- Migrate 'iframed masonry block' tests to Playwright. ([55016](https://github.com/WordPress/gutenberg/pull/55016))
- Migrate 'iframed multiple block stylesheets' tests to Playwright. ([55003](https://github.com/WordPress/gutenberg/pull/55003))
- Migrate keyboard-navigable-blocks end-to-end tests from puppeteer to playwright. ([54944](https://github.com/WordPress/gutenberg/pull/54944))
- Scripts: Properly use CommonJS for default Playwright configuration. ([54988](https://github.com/WordPress/gutenberg/pull/54988))
- Try fixing the flaky 'Toolbar roving tabindex' end-to-end test. ([54785](https://github.com/WordPress/gutenberg/pull/54785))
- end-to-end Tests: Revert temporary fixes. ([54865](https://github.com/WordPress/gutenberg/pull/54865))
- end-to-end Utils: Allow overriding username/password. ([53267](https://github.com/WordPress/gutenberg/pull/53267))

#### Build Tooling
- Add some @types packages as proper dependencies. ([50231](https://github.com/WordPress/gutenberg/pull/50231))
- Update the default JSX pragma to React instead of @wordpress/element. ([54494](https://github.com/WordPress/gutenberg/pull/54494))
- Upgrade wp-prettier to v3.0.3 (final). ([54775](https://github.com/WordPress/gutenberg/pull/54775))


### Security

#### Data Layer
- Replace turbo-combine-reducers with combineReducers from Redux. ([54606](https://github.com/WordPress/gutenberg/pull/54606))


### Various

- (chore) Revert the 16.7 RC2 release in order to release it again due to wrong changelog. ([54744](https://github.com/WordPress/gutenberg/pull/54744))

#### Design Tools
- Background image block support: Add tests, adjust injection logic slightly. ([54489](https://github.com/WordPress/gutenberg/pull/54489))
- Background support: Backport fix for undefined array key. ([54850](https://github.com/WordPress/gutenberg/pull/54850))

#### Typography
- Revert "Font Library: Avoid rendering font library ui outisde gutenberg plugin". ([54947](https://github.com/WordPress/gutenberg/pull/54947))

#### Commands
- Make the reset styles command consistent. ([54841](https://github.com/WordPress/gutenberg/pull/54841))

#### Patterns
- Use "Not synced" in place of "Standard" nomenclature for patterns. ([54839](https://github.com/WordPress/gutenberg/pull/54839))

#### Block Editor
- Simplify BlockPatternsSyncFilter with clearer labels and additional context. ([54838](https://github.com/WordPress/gutenberg/pull/54838))

#### Site Editor
- Use consistent capitalization for template parts in Site Editor constants. ([54709](https://github.com/WordPress/gutenberg/pull/54709))


"
`;

exports[`getContributorList verify that the contributors list is properly formatted 1`] = `
"

## Contributors

The following contributors merged PRs in this release:

@aaronrobertshaw @adamsilverstein @alexstine @andrewhayward @andrewserong @annezazu @anton-vlasenko @artemiomorales @aurooba @bangank36 @brookewp @c4rl0sbr4v0 @carolinan @chad1008 @ciampo @dcalhoun @derekblank @draganescu @ellatrix @fluiddot @fullofcaffeine @geriux @getdave @glendaviesnz @gziolo @jameskoster @jeryj @jsnajdr @juhi123 @kevin940726 @leemyongpakvn @madhusudhand @MaggieCabrera @Mamaduka @matiasbenedetto @michalczaplinski @mirka @mtias @mujuonly @ndiego @noahtallen @noisysocks @ntsekouras @oandregal @ockham @pbking @priethor @ramonjd @richtabor @scruffian @SiobhyB @spacedmonkey @stokesman @swissspidy @t-hamano @tellthemachines @tellyworth @them-es @torounit @tyxla @westonruter @WunderBart @youknowriad"
`;

exports[`getContributorProps verify that the contributors props are properly formatted 1`] = `
"## First-time contributors

The following PRs were merged by first-time contributors:

- @leemyongpakvn: Fix a broken MD link in callout. ([54772](https://github.com/WordPress/gutenberg/pull/54772))
"
`;
